### DISCUSSION TITLE
Server functions implementations

### DISCUSSION THREAD

**Discussion by mattaliev** [amannn/next-intl#1100] [General]
Hi, I have been using next-intl for a couple of days and you did such a great job! I really like it for it's simplicity to use and clean implementation. I am fairly new to next.js so I hope you don't mind if I ask you some questions about it, because I have a similar use case. More specifically, I want to know how do functions like getLocale() on the server work. I can see in the middleware, that you are setting a cookie and a header, but I couldn't find anywhere where you are reading it outside the middleware. 

Currently, I have a [store] route with a layout in it. And, I setup a context provider to use this store value in all the client components where it's needed, but for server components I am just prop drilling it down, which I am not a fan of. Ideally, I would have an asyncronous function getStore() that give me the store value. For this to work, I guess I have to set the cookie and the header in the middleware, but then how do I actually read it?

**Comment by amannn** [comment]
Hey, glad to hear that next-intl works well for you! 🙂 

There's some background on the locale passing here: [`unstable_setRequestLocale`](https://next-intl-docs.vercel.app/docs/getting-started/app-router/with-i18n-routing#setrequestlocale-background).

For next-intl, it seems to be the right choice with the header (and optionally the `unstable_setRequestLocale` workaround). If prop drilling isn't _too_ much of a hassle, I'd recommend to stick with that for your app though!

The ideal solution for this problem would be https://github.com/vercel/next.js/discussions/58862.
